src/mo_frontend/bi_match.ml

```diff
@@ -244,94 +331,178 @@ let bi_match_subs scope_opt tbs subs typ_opt =
   and bi_match_bind ts rel eq inst any tb1 tb2 =
     bi_match_typ rel eq inst any (open_ ts tb1.bound) (open_ ts tb2.bound)
 
-  and choose_under_constrained lb c ub =
-    match variance c with
-    | Variance.Covariant -> lb
-    | Variance.Contravariant -> ub
-    | Variance.Bivariant -> lb
-    | Variance.Invariant ->
-      raise (Bimatch (Format.asprintf
-        "implicit instantiation of type parameter %s is under-constrained with%a\nwhere%a\nso that explicit type instantiation is required"
-        (Cons.name c)
-        display_constraint (lb, c, ub)
-        display_rel (lb,"=/=",ub)))
-
-  and fail_over_constrained lb c ub =
-    raise (Bimatch (Format.asprintf
-      "implicit instantiation of type parameter %s is over-constrained with%a\nwhere%a\nso that no valid instantiation exists"
-      (Cons.name c)
-      display_constraint (lb, c, ub)
-      display_rel (lb, "</:", ub)))
-
-  and fail_open_bound c bd =
-    let c = Cons.name c in
-    raise (Bimatch (Format.asprintf
-      "type parameter %s has an open bound%a\nmentioning another type parameter, so that explicit type instantiation is required due to limitation of inference"
-      c (Lib.Format.display pp_typ) bd))
-
   in
-    let bds = List.map (fun tb -> open_ ts tb.bound) tbs in
-    List.iter2 (fun c bd -> if mentions bd cons then fail_open_bound c bd) cs bds;
-
-    let l = ConSet.fold (fun c l -> ConEnv.add c Non l) cons ConEnv.empty in
-    let u = ConSet.fold (fun c u -> ConEnv.add c (bound c) u) cons ConEnv.empty in
-
-    let l, u = match scope_opt, tbs with
-      | Some c, {sort = Scope; _}::tbs ->
-        let c0 = List.hd cs in
-        ConEnv.add c0 c l,
-        ConEnv.add c0 c u
-      | None, {sort = Scope; _}::tbs ->
-        raise (Bimatch "scope instantiation required but no scope available")
-      | _, _ ->
-        l,
-        u
+  bi_match_list bi_match_typ
+
+(** Solves the given constraints [ts1, ts2] in the given context [ctx].
+    Unused type variables can be deferred to the next round.
+    [deferred_typs] are types to appear in the constraints of the next round. Used to determine which type variables to defer.
+ *)
+let solve ctx (ts1, ts2) deferred_typs =
+  if debug then Debug.print_solve ctx (ts1, ts2);
+
+  (* Defer solving type variables that can be solved later. More constraints appear in the next round, let them influence as many variables as possible *)
+  let to_defer, defer_verify = if deferred_typs = [] then (ConSet.empty, false) else
```

Comment:
In the 1st round we should solve as few type variables as possible.
1. This avoids restricting the solution too early (more constraints are added in the next round)
2. and avoids inferring default bounds (like `Non`) for inputs to deferred functions (see [lambdas-open-types-in-deferred.mo](https://github.com/dfinity/motoko/pull/5180/files?diff=split&w=0#diff-828afbefe38428c6942a903c5467260bd51a28db56184c181f17417f7596281f)) that lead to confusing errors